### DISCUSSION TITLE
Preprocessing of 3D MRI such as motion correection, bias correction and noise removal

### DISCUSSION THREAD

**Discussion by Sunita2023** [dipy/dipy#3103] [Preprocessing]
Dear Dipy-developer,

How to perform motion correction, bias correction and noise removal of 3D MRI images (structural MRI: T1W, T2W, FLAIR, T1C) using DIPY?

We are recent user of DIPY. Most the tutorials given are for DWI.

Thank you so much in advance for your help.

Best regards,
Sunita

**Comment by Sunita2023** [comment]
Dear Jong,

I have a query.  
Could you please suggest on how to perform MRI Scanner noise removal and motion correction? 
I would highly appreciate your help.

Best regards,
Sunita

**Reply by pjsjongsung** [reply]
Hello Sunita,

DIPY (Diffusion Imaging for Python) is mostly focused on diffusion imaging, so we do not have a wide set of pre-processing for structural images.

That being said, I do believe, however, that noise removal (Something other than Patch2Self) and motion correction that we have for DWI should also work with structural MRI. Please let us know if you have some trouble making it work.

If you also have any suggestions and contributions you can make on the pre-processing methods in structural MRI, we would love to hear about it!

We are currently in the middle of a online workshop, so it might take a while for us to answer some questions. Apologies in advance. Just as a reminder, adding another comment sends us an alarm, so you do not have to close and reopen a post. If a specific individual is needed, pinging us this way ( @pjsjongsung  ) should also help.

**Reply by skoudoro** [reply]
Thanks @pjsjongsung,

I want to emphasize that DIPY is a community project. We love contribution from the community. All tutorials that you are reading or datasets are contributions from user (beginner or not) and we look forward to your participation in the process to extend the knowledge. if you need help to do so, we will be happy to help.

**Comment by Sunita2023** [comment]
Hi Jong,

Thank you so very much for explaining so clearly. I know that workshop is going on. 

We will try noise removal for DWI and structural MRI and let you know how it works for structural MRI. 

We are extracting FA and B0 from DWI. Then, generating streamline from FA image and registering it to the atlas (hcp_842). Using recobundle we are recognizing and refining the bundle of interest from the streamline.  When we visualize the bundle with the B0 image it is misaligned.  

Could you please help us no how to register B0 with the moved streamline so that it will align correctly with B0?

We will definitely participate and contribute in the knowledge transfer. 

Best regards,
Sunita

**Reply by skoudoro** [reply]
> Could you please help us no how to register B0 with the moved streamline so that it will align correctly with B0?

This tutorial seems to be enough close on what you are trying to achieve:

https://docs.dipy.org/stable/examples_built/registration/streamline_registration.html#sphx-glr-examples-built-registration-streamline-registration-py

**Comment by Sunita2023** [comment]
Hi Serge,

We followed the tutorial suggested by you however still having the same problem. 
Please find the link of the notebook and the input files. Kindly please guide us why the bundle is not aligned with the brain image even after registration. 

https://drive.google.com/drive/folders/1_y0AaWQRJM0nq4iPbDjG1WxplZRjC3h1?usp=sharing

We tried other methods too, but nothing is working out. 
Here we registered b0_dwi image with T2 MNI and tried seeing the bundle on B0_dwi.  However, it is misaligned. 

Your suggestion will be highly appreciated. 

Thank you so much.
Sunita

**Reply by skoudoro** [reply]
Hi @Sunita2023, 

Thank you for the folder. I will be able to look into it after March 24 since I am quite busy with the workshop attendees and follow up. I will keep you updated.

**Reply by pjsjongsung** [reply]
I have skimmed through your code and one of the cells

```
level_iters = [10, 10, 5]

final_warped_b0, mapping = syn_registration(
   mean_b0, t2_resliced_data, moving_affine=dwi_affine,
   static_affine=t2_resliced_affine, prealign=warped_b0_affine,
   level_iters=level_iters)
```

seems to use mean_b0 instead of warped_b0. Could this be one of the issues?

**Comment by Sunita2023** [comment]
Dear Jong and Serge,

Thanks Jong. In spite of your busy scheduled you looked at the code. 

For registration we are using affine and not syn registration to perform our task.  
We are trying to register dwi_b0 with t2 MNI so that the streamlines bundles and brain image take the same 3D space.  

The shape and voxel size of t2 MNI is (197, 233, 189) and (1 1 1) respectively. We resliced the t2 MNI to match the voxel size of dwi_b0 that is (2.5 2.5 2.5). After reslicing t2 MNI to the same voxel space as that of dwi_b0 registration is performed where dwi_b0 is registered to t2 MNI. The new shape of dwi_b0 after registration is (79, 93, 76) and voxel size is (2.5 2.5 2.5).  

To our surprise the initial slices in registered dwi_b0 are missing. When we are using this file for vtp construction instead of looking like a brain it is looking like a dome.  We want to show the streamline inside the the brain structure. When we are using original dwi_b0 there little misalignment of the streamline bundles. 

The main problem is how to align streamlines bundle and the 3D brain space?

We are just stuck at this point. I am sure your guidance will definitely help us to come out of this problem. 

Thanks again and wish you all the very best for the great success of the ongoing workshop event. 

Best regards,
Sunita

**Reply by skoudoro** [reply]
So, First, Thanks to @pjsjongsung.

I also skimmed the code and 

`reco_af_l = StatefulTractogram(r_recognized_bundle_l, atlas_header, Space.RASMM)` is wrong because wrong space. You can use `atlas_header` if you want to visualize it with the atlas. but if you want to use after with dwi, you need to use `dwi_img` instead. 

I can not test it now, like said above, only after March 24, but this is what popup to my eyes first.

**Reply by skoudoro** [reply]
Also, the `t2_resliced_data`  still have a skull, so the image registration is way off !!

please, check that `t2_resliced_data` and `mean_b0` have no skull and aligned after registration.

**Comment by Sunita2023** [comment]
I am truly grateful to Jong and Serge. I have certain deadline. Sorry to bother you. Any quick solution would work best. More extensively will ask again if any unsolved question is there after March 24th.  Thanks a lot!

With the addition of rrb = StatefulTractogram(r_recognized_bundle, dwi_img, Space.RASMM) and
 save_trk(rrb, vtpfilepath, bbox_valid_check=False) it seems to be working. However not fully.

 For example in the attached image, streamlines are coming out of dwi_b0 image. How to fix this? It is the same DWI image then how it is coming out of dwi_b0?
![bundle_and_brain](https://github.com/dipy/dipy/assets/112482239/47f34f83-a059-4da5-955b-46af81f4e832)

When I am saving trk file using the code "save_trk(rrb, vtpfilepath, bbox_valid_check=True)" I am getting the error message ERROR:StatefulTractogram:Voxel space values higher than dimensions.

There is one more question. I could not generate PC.trk bundle using recognize and refine. It is giving following error message. AttributeError: 'ClusterMapCentroid' object has no attribute 'get_clusters'
Reference PC.trk bundle is there still I am not able to generate it. Why is so?

What you would advice, generating bundle via extraction or from the seed? CSD deterministic or CSD probabilistic which one is best to use? At this point we are using peak_indices = peaks_from_model(
    model=dti_model, data=dwi_data, sphere=sphere, relative_peak_threshold=.2,
    min_separation_angle=25, mask=dwi_mask, npeaks=2, parallel=True, num_processes=4)

My sincere thanks to both of you.

Best regards,
Sunita

**Reply by skoudoro** [reply]
> For example in the attached image, streamlines are coming out of dwi_b0 image. How to fix this? It is the same DWI image then how it is coming out of dwi_b0?

Can you tell us how this image has been generated?  also which tools did you to visualize ? How this tool handle affine and different space? I do not know if the issue is coming from the code or the visualization tool.

> When I am saving trk file using the code "save_trk(rrb, vtpfilepath, bbox_valid_check=True)" I am getting the error message ERROR:StatefulTractogram:Voxel space values higher than dimensions.

it can make sense with the image above.  It look like an affine ( or space) issue.

> There is one more question. I could not generate PC.trk bundle using recognize and refine. It is giving following error message. AttributeError: 'ClusterMapCentroid' object has no attribute 'get_clusters'
Reference PC.trk bundle is there still I am not able to generate it. Why is so?

This error means that the volume is empty as some point so it could not find it.  Not sure why you get this error and it will be interesting to introspect it with your data + script. Maybe this bundle just do not exist on your data.

> generating bundle via extraction or from the seed

I believe both are needed to be the closest as possible but any other opinion are welcomed !

> CSD deterministic or CSD probabilistic which one is best to use?

Unfortunately, it is not binary, Depends on your application,  your needs , your data etc... there is always a tradeoff:  performance vs  precision.


it will help if you could present briefly yourself, the project, your goal and your data (classic, template , specific pathology, adult, baby, etc...)

**Reply by skoudoro** [reply]
Also, in the image above, it seems there is a flip between tracks and brain, so, it is an affine/space issue for sure (they are not in the same coordinate system)

**Reply by Sunita2023** [reply]
Thank you so much Serge for responding to each of my query.  I am sincerely grateful to you. 

The image is generated was from dwi_b0 nifti file which was converted to vtp. Similarly streamlines are also converted to vtp. Nifti file dwi_b0 uses the same affine as that of DWI from which streamlines are generated. But after registration of the streamlines with the atlas the affine changes from that of dwi_b0. How can we change the affine of dwi_b0 like that of streamline? 

I read somewhere that for specific pathology or clinical condition  CSD deterministic is better. What is your comment on this?

Thanks again for your guidance. 
Best regards
Sunita
